### ISSUE TITLE
Add c++17 support

### ISSUE THREAD

**Issue by chuggafan** [LADSoft/OrangeC#147]
This is a request for a "long term goal" of adding c++17 support, with a lot of new features and useful items that would allow one to utilize features such as the "filesystem" library that llvm and Microsoft have, meaning you can use it for file manipulation easily as utilize features such as ```[[nodiscard]]``` which gives users a warning that the value returned from a function is usually used, as well as implement new, useful keywords such as ```byte``` (which is not a keyword in previous versions of c++) which could be used for explicitly wanting a 8 bit byte, as well as other features, such as nested namespaces, where one can do:
```cpp

namespace A::B::C {

}
```
instead of:
```cpp

namespace A {
    namespace B {
        namespace C {
}}}
```

This is a LONG TERM GOAL, so I hope it can be implemented eventually, the other thing I would like to ask in this issue is the implementation of feature test macros for components that exist within the compiler, the macros that could be used to detect certain features are given [here](http://en.cppreference.com/w/User:D41D8CD98F/feature_testing_macros#C.2B.2B14) for showing what values each macro should hold if it exists at all

I'd also like to again, express my thanks for all of the hard work you've done, I am currently still working on that threading testing, but I will also look into attempting to take what you've done with your c11 threading header and convert it to being able to be used for the c++ threading section as well

**Comment by LADSoft** [comment]
yeah I do have a long term goal of implementing C++2017; I had just completed C++2014 last year though and was taking a break lol!   But when I looked at the list of features in 2017 it didn't seem *too* bad given that I can reuse the llvm library for the really hard stuff...

Thank you for your appreciation, it means a lot to me :)

**Comment by LADSoft** [comment]
BTW I was just going to leave the issue about the multithreading support open and see if you would find time for it at some point...   it isn't like i don't have a ton of other things to do lol!

**Comment by LADSoft** [comment]
it looks like fold expressions and class template auto deduction are the most difficult things to implement in this version of the standard; everything else should proceed easily enough.

**Comment by LADSoft** [comment]
so i've started on c++17 support now; im thinking in terms of trying to get a release with this and C2X support sometime around christmas.   It will also be time to bump the compiler to version 7 when we get this done...

**Comment by LADSoft** [comment]
the major things left are copy elision and CTAD...   there are a couple of minor things as well.

**Comment by LADSoft** [comment]
I have one more feature to add: lambda capture of *this.    It shouldn't be a big deal...   after that I'll seriously try compiling some stuff in c++17 mode to see how it handles the libraries....

**Comment by LADSoft** [comment]
so im back to working on this - added lambda capture of *this and fixed some bugs with lambda functions.    I'm in the process of compiling against the c++17 version of libcpp, have fixed several bugs but right now the issue is that c++14 variable templates don't quite work properly.   Hope to have this fixed within the next few days...

after I get this working I will dive into all the recently added items in the backlog/discussions.

**Comment by LADSoft** [comment]
so making progress.   Over the last few weeks ive rewritten the constexpression handling to handle various use cases that came up with running the lbcxx tests for c++17....  I'm slowly working through the tests.   One crashes during compile, A few others still don't compile, a lot of the executables throw asserts when they are run, and I guess a couple of executables are stuck in infinite loops.    But there are less than 100 open issues at this point...   I'll be working through them over the next few weeks.    

Once we get this done I have a few minor things to investigate that may improve the speed of the compiler, and once I'm through that we will release c++17 version.

**Comment by LADSoft** [comment]
ive been off in various distractions the last few weeks and haven't gotten back to this... yet...   i did finally look into the appveyor build problem though, where the appveyor builds have been failing since march or so.   Compiles are way longer than in the last previous build on appveyor, last October and they run out of time.   whereas it had been stable for a while before that...

After a bunch of research it seems like the root issue is that the compiler slows down dramatically, when C++17 mode is selected and intense use of LIBCXX headers comes into play..   Possibly a big part of that being, something being wrong with the constexpr processing.    I'm running some tests on appveyor over the next couple of days to see if I can confirm those observations and then will work on getting the actual problem fixed....

I'm also working toward making a change that will cut down a lot of duplicate work the compiler is doing when it compiles types.   If I'm right we'll get some sort of speedup out of that as well...

**Comment by LADSoft** [comment]
so after going off on tangent after tangent for several days I finally pulled the plug on tangents and now I'm working on the slow compile speeds again.

i got from testing over a couple of nights that MSVC compiles the compiler 15% slower now.   And when compiling the compiler with a version of itself compiled with MSVC it is also about 15% slower.

I trolled through a bunch of old commits doing comparisons of compile times, and then honed in on two interesting points, first setting the version of the compiler as C++17 for purposes of header evaluations added to the compile time, and second at one point just adding an external declaration to a header slowed the compiles down by several hundred milliseconds.

I can't fix the first part directly.

But I think I figured out, sometimes if I can defer instantiation of types until they are actually used it might help compile speeds...

as a simple example:

```
extern std::map<int, double> aa;
```

the type doesn't NEED to be instantiated unless aa is actually used...   so the evaluation could be deferred or maybe not even done at all.    And since instantiating it then drags in a lot of other types that ALSO are instantiated even though they don't have to be, it could take significant time just to instantiate that one type.

there are other similar situations where types don't need to be instantiated unless some variable involved is being defined or referenced......

I got some numbers, and if I'm at all in the ballpark defering type evaluations could result in a substantial decrease of compile times for the files in the occparse directory...

so that is what I'm working on this week, exploring how far I can go in deferring type instantiations....

**Comment by LADSoft** [comment]
ok so after a lot of head-scratching i finally got to the point of being able to do benchmarks on the change I made (discussed in last comments)  and I'm quite happy.    I was seeing about a 30% decrease in compile times in basic tests, and that was born out when I compiled OCC with MSVC and then used that version to compile it with itself.   Just compiling the compiler with one omake thread, not doing a full build of the system I got about a 31% decrease in compile time over the last version of the compiler right before I made the change.    

So  I did the build script as well.   That does multiple threads and compiling occ with itself multiple times and builds the RTL a couple of times and builds all the test programs as well as makes zip files; I got about a 23% decrease in compile times there.    So I'm hoping that when I move this to main and let appveyor have at it we'll be back within the compile time limits for appveyor...

Even though i have a lot more i want to do in this vein, I'm gonna quit with the side projects for a while and concentrate on getting c++17 ready for release.    There are four parts to this:

1) fix a couple of minor bugs
2) #592 add file system header back in
3) clean up various issues where the compiler won't compile programs that have CMAKE scripts
4) finish the libcxx tests...

hoping to have the first three this weekend then it will possibly take a few weeks to finish up the libcxx tests....   after that it will be time for a release.     Once I release I'll go back to the CMAKE development team and get the scripts updated;   we'll have added C2x and c++17 in this release.

**Comment by LADSoft** [comment]
so the current progress is that after another minor rewrite I've finally got constexpression worked out for the filesystem headers...    hopefully I can finish up with filesystem over the next week or so.

**Comment by LADSoft** [comment]
the other thing is, the rewrite I did, deferring type instantiations, was successful.    Builds are a little faster again, and we got an appveyor build to compile orangec within the time limits now.

**Comment by LADSoft** [comment]
I got the c++17 filesystem runtime to compile last night...   couple of link errors left...    and then a lot of linux-specific function calls are being used.   This weekend I'll look and see if there is a version more suitable for windows but if not I guess there is are bunch of library calls to implement before I can call this done...   while I'm waiting for the weekend I'm going to get the c++17 variant stuff compiling them see if I can figure out why the tests aren't compiling any more...

**Comment by chuggafan** [comment]
For the filesystem, could it be that we're on too old of a libcxx for
<filesystem> to be implemented properly on win32? I know we're on a
decently old version of libcxx (the commit for the <filesystem> header
lists libcxx 10) and llvm recently released 18.1.8 (2 releases per year, so
it's been about ~4 years).
I am seeing specific MSVC stuff for <filesystem>... perhaps it's not
completely using it? I notice directory_iterator.cpp has specific FIXMEs
for windows on our version, so...

**Comment by LADSoft** [comment]
yeah maybe i should look at upgrading.    That is always a major production though....   and i feel like this has been dragging on and on and on for the last 10 or 12 months lol...    maybe this weekend I'll get the latest and at least see if I can figure out what the scope of work for upgrading will be :)

**Comment by chuggafan** [comment]
My look into it says that clang only really likes to have the most
recent compilers being supported, but it currently attempts to support
C++11 or greater compilers.
WIth MSVC support for filesystem, it also is fairly clear that it DOES
NOT by default support MSVC in some of the releases, I'd have to check
the very latest but 13.0.0 mentions:

```
std::filesystem is now feature complete for the Windows platform using
MinGW. MSVC isn’t supported since it lacks 128-bit integer support.
```
https://releases.llvm.org/13.0.0/projects/libcxx/docs/ReleaseNotes.html

A quick inspection of the latest main branch of libcxx might indicate
that it doesn't *necessarily* still require that, it's still
mentioning 128 bit necessity AND by default disabling <filesystem> on
MSVC, but I have no idea why... if we are desperate, we could
theoretically steal the <filesystem> impl from the MSVC stl since that
went open source several years ago (under a similar license,
Apache-2.0 (with different exceptions, to be fair), no less!)
Of course, all of that in of itself is its own headache.

**Comment by LADSoft** [comment]
so i can implement 128 bit support if it needs it lol...   i'd prefer not to mix and match microsoft headers with libcxx headers if I can get away with it.   I know that the version of libcxx filesystem code we've got attempts to work around the 128 bit issue where it is relevant but I don't know maybe they took that out again...

**Comment by LADSoft** [comment]
i decided to try out libcxx 17 as that is the last 'finished' version (18 is in progress).    I decided to try my hand at compiling a vector as that is one of the simpler headers...

The main problems seem to be that there are new intrinsics I have to implement (for example __is_same and there are at least two others) and something is broken with __has_include_next as it isn't finding the core C headers....   I'm wondering if the fact that the library now has subdirectories in the include directory has anything to do with that.    there are some other minor bugs as well...   this isn't seeming too bad at the moment so i'll keep going for now.

**Comment by LADSoft** [comment]
i implemented enough bug fixes to get vector.h in the libcxx 17 release to build and rolled the fixes back into the main line.   That cleaned up the filesystem support a little too...    Along the way I decided to speed compiles up by supporting '#pragma GCC system_header' properly, and found out that occ doesn't support CTAD properly...   good enough for implicit CTAD but I need to add the explicit support I guess.   I also found out that occ was generating unnecessary destructor calls for the iterators in all the for-range instantiations, so I cleaned that up a bit.   So it was a good use of time, taking this detour through libcxx17.

At this point I decided to move back to libcxx 10 though rather than continue with libcxx17, mostly I want a release sooner rather than later and that is just going to be more delays while I work out details.   I fixed some bugs compiling the filesystem support and implemented some missing unix functions, now I have one more bug to fix then the filesystem support will be integrated and I can do basic tests on that...   hope to get that done tonite.   And finish the CTAD over the next few days.   Then I'm done with the planned changes for this release.   Next I'm going to compile a bunch of test projects I've gathered over the last couple years as well as fix the libcxx tests again.  Then we will have a release hopefully within the next couple of months.

**Comment by chuggafan** [comment]
Awesome, I've been trying the build with the new support, and I'm getting
errors with <variant> where std::is_copy_constructible_v<_Tp> cannot
generate a template specialization, I'm assuming that it's fundamentally an
issue with some variable template issue as  that's mentioned as a specific
callout for the block of code in particular there
(_LIBCPP_HAS_NO_VARIABLE_TEMPLATES)
I'm still having massive issues with debugging why my code was breaking
with regards to shared_ptr instantiations, my best guess is everything I
previously thought was wrong. When I do try to take the journey via
manually setting breakpoints with hitcounts & conditions, the error doesn't
appear until much later (confirmed by going up one level and the error not
being encountered).
Somehow we're getting *some* broken behavior, if I had to guess the worst I
somehow forgot a return statement somewhere and that could cause the issue,
but I don't think that's it.
If you want your crack at the issue, I've pushed my current changes, but
outside of that issue there seems to be a template unpacking issue with ...
specifiers:
Error(413)    c:\ORANGEC\include\variant(1165):  Cannot generate template
specialization from 'std::is_copy_constructible_v<_Tp>'
note:         c:\ORANGEC\include\variant(1164):  Referenced in
instantiation of 'std::variant<shared_ptr<Token>, unsigned long long>'
note:         ppEmbed.h(31):  In template instantiation started here
Error(232)    ppEmbed.cpp(39):  Could not find a match for
'std::vector<variant<shared_ptr<Token>, unsigned long long>,
allocator<variant<shared_ptr<Token>, unsigned long
long>>>::push_back(shared_ptr<Token>)'
Error(232)    ppEmbed.cpp(47):  Could not find a match for
'std::vector<variant<shared_ptr<Token>, unsigned long long>,
allocator<variant<shared_ptr<Token>, unsigned long long>>>::push_back(
const unsigned long long)'
Error(232)    ppEmbed.cpp(118):  Could not find a match for
'std::vector<variant<shared_ptr<Token>, unsigned long long>,
allocator<variant<shared_ptr<Token>, unsigned long
long>>>::insert(__wrap_iter<variant<shared_ptr<Token>, unsigned long
long>*>, __wrap_iter<unsigned long long*>, __wrap_iter<unsigned long
long*>)'
Error(426)    c:\ORANGEC\include\variant(1589):  Pack ... specifier must be
used on function argument
note:         c:\ORANGEC\include\variant(1587):  Referenced in
instantiation of 'std::visit<visitor_struct&, variant<shared_ptr<Token>,
unsigned long long>&>(visitor_struct&, variant<shared_ptr<Token>, unsigned
long long>&)'
note:         ppEmbed.cpp(164):  Referenced in instantiation of
'tokens_to_inject(vector<variant<shared_ptr<Token>, unsigned long long>,
allocator<variant<shared_ptr<Token>, unsigned long long>>>)'
note:         ppEmbed.cpp(175):  In template instantiation started here
Error( 18)    c:\ORANGEC\include\variant(561):  Undefined symbol '__impl'
note:         c:\ORANGEC\include\variant(558):  Referenced in instantiation
of
'std::__variant_detail::__visitation::__variant::__visit_alt<__value_visitor<visitor_struct&>,
variant<shared_ptr<Token>, unsigned long
long>&>(__value_visitor<visitor_struct&>&&, variant<shared_ptr<Token>,
unsigned long long>&)'
note:         c:\ORANGEC\include\variant(576):  Referenced in instantiation
of
'std::__variant_detail::__visitation::__variant::__visit_value<visitor_struct&,
variant<shared_ptr<Token>, unsigned long long>&>(visitor_struct&,
variant<shared_ptr<Token>, unsigned long long>&)'
note:         c:\ORANGEC\include\variant(1587):  Referenced in
instantiation of 'std::visit<visitor_struct&, variant<shared_ptr<Token>,
unsigned long long>&>(visitor_struct&, variant<shared_ptr<Token>, unsigned
long long>&)'
note:         ppEmbed.cpp(164):  Referenced in instantiation of
'tokens_to_inject(vector<variant<shared_ptr<Token>, unsigned long long>,
allocator<variant<shared_ptr<Token>, unsigned long long>>>)'
note:         ppEmbed.cpp(175):  In template instantiation started here
Error(510)    c:\ORANGEC\include\tuple(1073):  static_assert generated:
type not in empty type list
note:         c:\ORANGEC\include\tuple(1080):  Referenced in instantiation
of 'std::__find_exactly_one_t<unsigned long long, shared_ptr<Token>,
unsigned long long>'
note:         c:\ORANGEC\include\variant(1371):  Referenced in
instantiation of 'std::holds_alternative<unsigned long long,
shared_ptr<Token>, unsigned long long>( const variant<shared_ptr<Token>,
unsigned long long>&)'
note:         ppEmbed.cpp(164):  Referenced in instantiation of
'tokens_to_inject(vector<variant<shared_ptr<Token>, unsigned long long>,
allocator<variant<shared_ptr<Token>, unsigned long long>>>)'
note:         ppEmbed.cpp(176):  In template instantiation started here
7 Errors

I guess it really doesn't like <variant> after all, I might switch away
from <variant> just to make that headache go away in the meantime but the
std::shared_ptr error is absolutely beyond me and anything I was doing with
debugging at the time, something about it is just *trying* to evade
detection and stepping through a bunch of code that isn't demonstrating the
issue isn't helping me.

**Comment by chuggafan** [comment]
Whoops, wrong thread, I meant to put the speil mostly in the PR thread for
my current PR.
Oh well.

**Comment by LADSoft** [comment]
thanks for keeping me informed of your ongoing problems with this.    I will probably look at it soon (possibly as early as tomorrow) as  currently there are a lot random build errors going on on appveyor and github...   I've duplicated a couple of them but rarely....   this is as good a place as any to start trying to to dig into this.

FWIW I got a simple filesystem test program to create a directory, and show a directory listing.   I did have to monkey with the libcxx10 filesystem stuff as there were some bugs especially in the WIN32 support.

**Comment by LADSoft** [comment]
@chuggafan 

so i was refactoring a bunch of stuff and kinda stumbled onto the kind of problem that illicits random behavior...     the refactor forced it to have bad results every time.    I will put a new version up later to see where we are with that...

I reworked the ctad to make it better for implicit deduction guides, now I have to work on explicit deduction guides.

Meanwhile I was able to easily reproduce the variant problem...    will get to fixing it after working on the explicit deduction guides.    If you want to poke around in the meantime I realized you may not know it, but, for parser problems if you set a breakpoint where the error is displayed (currently around line 470 of ccerr.cpp) you can often hone in directly on where any problem lies without a whole of extra effort

**Comment by LADSoft** [comment]
the variant header poses a bunch of new problems beyond the obvious one that started this off....  a lot of it is related to using packed template arguments with templated variables but there is other stuff as well...    I am slowly working through the issues.   I'm hoping to finish it this weekend but we will see...

**Comment by LADSoft** [comment]
im still working on the variant header, it is one issue after another.    I got to a certain point then realized I have to rewrite how unpacking packed template parameters works....    so that is what is going on right now.

**Comment by LADSoft** [comment]
so i about have the variant header worked out, my test program compiles although it doesn't run...   should have that fixed soon....

**Comment by LADSoft** [comment]
so the variant test program I have runs reliably now.    There are some problems with constexpression functions and structures being evaluated at run time but it is at least functional...   the test program has been added to the tests directory.

that ends the development activities for this release.   I'm going to do various cleanup this weekend...   and there will be a bit of a delay because I'm electing to do some extensive testing before release....

**Comment by GitMensch** [comment]
Which programs do you plan to test this on? Will those get into the internal test suite or is it more reasonable to have specific issues (outlined here) in a self-written program and only do an integration test on some application that uses c++17 manually?

**Comment by LADSoft** [comment]
so these are all the cmake projects that got referenced in other issues.   I've already worked on some of them but basically want to revisit  those and work on the ones I haven't.

Some of these aren't going to run due to missing runtime functions but for those I just want to make sure the build runs through other than the link stage.

asl
covscript
duilib
gravity
katascript
simplecpp

I don't know how suitable it is to add them to the test suite although I would like to where applicable.   If I can get APPVEYOR and GH to have an appropriate version of cmake...
